R/TealAppDriver.R

```diff
@@ -0,0 +1,425 @@
+        info = "Shiny error is observed"
+      )
+    },
+    #' @description
+    #' Check if the app has no validation errors. This checks for global shiny validation errors.
+    expect_no_validation_error = function() {
+      testthat::expect_null(
+        self$get_html(".shiny-output-error-validation"),
+        info = "No validation error is observed"
+      )
+    },
+    #' @description
+    #' Check if the app has validation errors. This checks for global shiny validation errors.
+    expect_validation_error = function() {
+      testthat::expect_false(
+        is.null(self$get_html(".shiny-output-error-validation")),
+        info = "Validation error is not observed"
+      )
+    },
+    #' @description
+    #' Set the input in the `teal` app.
+    #'
+    #' @param input_id (character) The shiny input id with it's complete name space.
+    #' @param value The value to set the input to.
+    #' @param ... Additional arguments to be passed to `shinytest2::AppDriver$set_inputs`
+    #'
+    #' @return The `TealAppDriver` object invisibly.
+    set_input = function(input_id, value, ...) {
+      do.call(
+        self$set_inputs,
+        c(setNames(list(value), input_id), list(...))
+      )
+      invisible(self)
+    },
+    #' @description
+    #' Navigate the teal tabs in the `teal` app.
+    #'
+    #' @param tabs (character) Labels of tabs to navigate to. The order of the tabs is important,
+    #' and it should start with the most parent level tab.
+    #' Note: In case the teal tab group has duplicate names, the first tab will be selected,
+    #' If you wish to select the second tab with the same name, use the suffix "_1".
+    #' If you wish to select the third tab with the same name, use the suffix "_2" and so on.
+    #'
+    #' @return The `TealAppDriver` object invisibly.
+    navigate_teal_tab = function(tabs) {
+      for (tab in tabs) {
+        self$set_input(
+          sprintf("teal-main_ui-%s-active_tab", private$modules$label),
+          get_unique_labels(tab),
+          wait_ = FALSE
+        )
+        root <- sprintf("%s-%s", private$modules$label, get_unique_labels(tab))
+      }
+      self$wait_for_idle(timeout = private$idle_timeout)
+      private$set_active_ns()
+      invisible(self)
+    },
+    #' @description
+    #' Get the active shiny name space for the Module content and the Filter panel.
+    #' Note that in the case of the filter panel, the name space is constant when it is not module specific.
+    #' However, module specific filter panel will have the name space linked with the module name space.
+    #'
+    #' @param component (character) The component to get the active name space for.
+    #' Currently supported components are "module", "filter_panel", and "filter_manager".
+    #'
+    #' @return The active shiny name space of the component.
+    get_active_ns = function(component) {
+      switch(component,
```

Comment:
Or let's discuss alternative to trigger this call :-)

Reply:
I made changes to the `get_active_ns` as per the previous comments in the commit https://github.com/insightsengineering/teal/pull/1127/commits/22df0a2cccb1832c1abb32e8b8022eae31a6eb10

Reply:
The problem still persists with teal_data_modules

What do you think about changing `active_ns` to an active binding and using it instead of `private$ns` directly?

```r
    #' @description
    #' Get the active shiny name space for interacting with the module content.
    #'
    #' @return (`string`) The active shiny name space of the component.
    active_module_ns = function() {
      self$active_ns$module
    },
    #' @description
    #' Get the active shiny name space for interacting with the filter panel.
    #'
    #' @return (`string`) The active shiny name space of the component.
    active_filters_ns = function() {
      self$active_ns$filter_panel
    },
    #' @description
    #' Get the active shiny name space for interacting with the filter panel.
    #'
    #' @return (`string`) The active shiny name space of the component.
    filter_manager_ns = function() {
      self$active_ns$filter_manager
    },
   # ...
  ),
  active = list(
    active_ns = function() {
      if (all(vapply(private$ns, length, integer(1)) == 0)) {
        private$set_active_ns()
      }
      private$ns
    }
  ),
```

Reply:
Can you try this. Making the `set_active_ns` public and calling it once the teal app is initialised after the DDL.

Here's an example app after you make it public
```r
tdm <- teal_data_module(
  ui = function(id) {
    ns <- NS(id)
    actionButton(ns("submit"), label = "Load data")
  },
  server = function(id) {
    moduleServer(id, function(input, output, session) {
      eventReactive(input$submit, {
        data <- within(
          teal_data(),
          {
            dataset1 <- iris
            dataset2 <- mtcars
          }
        )
        datanames(data) <- c("dataset1", "dataset2")
        data
      })
    })
  }
)

app <- TealAppDriver$new(
  data = tdm,
  modules = example_module(label = "Example Module")
)
app$view()

app$active_ns()
app$click("teal_data_module-submit")
app$set_active_ns()
app$active_ns()

app$stop()
```

Reply:
Do you think `set_active_ns` should be public? If so then why call it explicitly here only and not while initialization or after navigation?

This might protect the `active_ns` from other (future) cases when it's not initialized.

Reply:
Setting the name space when not initially set 29cbca7b0f4ab904b872f8e0313758c40f08b7d5. The only reason why this failed was because the app was not initiated.